lib/shopify_api/api_version.rb

```diff
@@ -109,5 +110,15 @@ def construct_graphql_path
         construct_api_path('graphql.json')
       end
     end
+
+    class NullVersion
+      def self.nil?
```

Comment:
Null object does not have to implement `nil?`

Instead implement a null version of all the methods on ApiVersion.

My suggestion would be 
```ruby
class NoVersionConfigured < StandardError; end

class NullVersion < ApiVersion
  def initialize
    @version_name = "no-version"
    @numeric_version = 0
  end

  def stable? 
     false
  end
  
  def construct_api_path(path)
        raise ApiVersionNotSetError, "You must set ShopifyAPI::Base.api_version before making a request."
  end
  
  def construct_graphql_path
      raise ApiVersionNotSetError, "You must set ShopifyAPI::Base.api_version before making a request."
  end
end
```

Reply:
If someone was checking to see if the version was nil, but it was actually NullVersion, I figured it fulfills the same meaning in that its not a version. Its just a fancier NilClass

I also think its probably misleading to have a null version respond to message other than raising an error. `ShopifyAPI::Base.api_version.stable? # => false` implies there is a version set, its just not stable (:unstable perhaps). I did consider implementing every method and having it raise, but method_missing seemed like an easier choice.

Reply:
We could have it raise on all methods that matter.  The main goal of the Null Object pattern is to remove nil checks and always have an implementation of the interface.  Responding `true` to the `nil?` method would encourage the thing that we are trying to remove.

Reply:
The behavior we're trying to remove sending messages ApiVersion messages to NilObject where we expect an instance of ApiVersion and raising a no_method error. We don't want to provide any default behavior, we just want to replace any methods that ApiVersion has raise with a more meaningful message and avoid adding conditionals to handle the NoMethodError. 

If we don't redefine `nil?` to `true` (and in hindsight, empty? and present?), then those values default to making the object truthy. Developers will still check `if ShopifyAPI::ApiVersion.api_version.nil?` but the code will instead now blowup at an expected location opposed to where they expected it to possibly be nil.

Reply:
So there are two different ways of conceptualizing the NullObject pattern: as a "better null", which just has nicer errors or other properties, or as a specific non-null instance of the object which happens to get used when you might otherwise return null. Without getting too into the weeds of OO design, either will work for this particular case, though the latter (as Alex has suggested) is usually more common and probably what I'd go with for convention.